sw/device/silicon_creator/lib/drivers/otp.c

```diff
@@ -6,28 +6,39 @@
 
 #include <stddef.h>
 
-#include "sw/device/lib/base/bitfield.h"
-#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/silicon_creator/lib/base/abs_mmio.h"
 #include "sw/device/silicon_creator/lib/error.h"
 
-#include "otp_ctrl_regs.h"
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "otp_ctrl_regs.h"  // Generated.
 
-uint32_t otp_read32(const otp_t *otp, uint32_t address) {
-  return mmio_region_read32(otp->base_addr,
-                            OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET + address);
+uint32_t otp_read32(uint32_t address) {
+  return abs_mmio_read32(TOP_EARLGREY_OTP_CTRL_BASE_ADDR +
+                         OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET + address);
 }
 
-uint64_t otp_read64(const otp_t *otp, uint32_t address) {
+uint64_t otp_read64(uint32_t address) {
   ptrdiff_t reg_offset = OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET + address;
-  uint64_t value =
-      mmio_region_read32(otp->base_addr, reg_offset + sizeof(uint32_t));
+  uint64_t value = abs_mmio_read32(TOP_EARLGREY_OTP_CTRL_BASE_ADDR +
+                                   reg_offset + sizeof(uint32_t));
   value <<= 32;
-  value |= mmio_region_read32(otp->base_addr, reg_offset);
+  value |= abs_mmio_read32(TOP_EARLGREY_OTP_CTRL_BASE_ADDR + reg_offset);
 
   return value;
 }
 
-void otp_read(const otp_t *otp, uint32_t address, void *data, size_t len) {
+rom_error_t otp_read(uint32_t address, void *data, size_t len) {
+  if (address % alignof(uint32_t) != 0 || len % alignof(uint32_t) != 0) {
+    return kErrorOtpBadAlignment;
+  }
+
   ptrdiff_t reg_offset = OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET + address;
-  mmio_region_memcpy_from_mmio32(otp->base_addr, reg_offset, data, len);
+  for (size_t i = 0; i < len; i += sizeof(uint32_t)) {
+    uint32_t word =
+        abs_mmio_read32(TOP_EARLGREY_OTP_CTRL_BASE_ADDR + reg_offset + i);
+    memcpy(data + i, &word, sizeof(uint32_t));
```

Comment:
for sensitive operations, it is preferred to use naturally aligned memory access, so if we can, lets switch to a 32bit aligned output. This will also simplify integration with sec_mmio later on. 

@cfrantz is this recommendation ok from an integration perspective?

Reply:
There are some cases where we want to do byte-level access on values read from OTP, but in all such cases, we still want to read out the entire OTP word.  I think it makes sense to do the OTP reads in terms of 32-bit words and leave interpreting the values to the caller.

Reply:
To clarify my position on this: since it only makes sense to read word-sized values out of OTP, this function should only accept an out-pointer to a word-sized entity:

```
rom_error_t otp_read(uint32_t address, uint32_t *data, size_t num_words)
```

Reply:
Please address this comment (otherwise, everything lgtm).